### PR TITLE
[2/N] Define dataclasses for progress tracking

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Continues #6603
Part of #6429 

The following google document documents the purpose for each of these attributes

https://docs.google.com/document/d/1U73b4uFML2MNdObh6MVsBpFDrgLUOU2Sd_x0_DVk-Bo/edit?usp=sharing

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 ### Added
 
 
+- Added dataclasses for progress tracking ([#6603](https://github.com/PyTorchLightning/pytorch-lightning/pull/6603))
```
Comment (carmocca) [inline CHANGELOG.md:13]:
```suggestion
- Added dataclasses for progress tracking ([#7574 ](https://github.com/PyTorchLightning/pytorch-lightning/pull/7574))
```

File: pytorch_lightning/trainer/progress.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    """
+    Basic dataclass to track event progress.
+
+    Args:
+        ready: Intended to track the number of events ready to start.
+        started: Intended to be incremented after ``on_*_start`` completes.
+        processed: Intended to be incremented after the event is processed.
+        completed: Intended to be incremented after ``on_*_end`` completes.
+
+    Attributes set to ``None`` are treated as unused and are restricted.
```
Comment (carmocca) [inline pytorch_lightning/trainer/progress.py:29]: @ananthsub We now hide "unused" attributes, only used in `OptimizerProgress`

File: pytorch_lightning/trainer/progress.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    optimization: OptimizationProgress = field(default_factory=OptimizationProgress)
+
+
+@dataclass
+class TrainLoopProgress(LoopProgress):
+    epoch: TrainProgress = field(default_factory=TrainProgress)
+
+    def reset_on_epoch(self) -> None:
+        # override to avoid resetting `epoch.current`
+        self.batch.current.reset()
```
Comment (carmocca) [inline pytorch_lightning/trainer/progress.py:172]: @ananthsub We need to override this method in the case of the training loop as `epoch.current` should track the "number of epochs this fit() call"

File: tests/trainer/test_progress.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    with pytest.raises(AttributeError, match="object has no attribute 'non_existing_attr'"):
+        p.non_existing_attr  # noqa
+    # can set new attr
+    p.non_existing_attr = 10
+    # can't write unused attr
+    with pytest.raises(AttributeError, match="'completed' attribute is meant to be unused"):
+        p.completed = 10
+    with pytest.raises(TypeError, match="unsupported operand type"):
+        # default python error, would need to override `__getattribute__`
+        # but we want to allow reading the `None value`
```
Comment (carmocca) [inline tests/trainer/test_progress.py:34]:
```suggestion
        # but we want to allow reading the `None` value
```

File: pytorch_lightning/trainer/progress.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        if self.processed is not None:
+            self.processed = 0
+        if self.completed is not None:
+            self.completed = 0
+
+    def __setattr__(self, key: str, value: int) -> None:
+        if getattr(self, key, 0) is None:
+            raise AttributeError(f"The '{key}' attribute is meant to be unused")
+        return super().__setattr__(key, value)
+
```
Comment (ananthsub) [inline pytorch_lightning/trainer/progress.py:50]: this means at construction we need to null out the fields we want to mark as unused. for users, indicating this behavior change with the API based on types feels hidden. this also means if anyone extends this class and really wants an optional field which is initially None but later tracked for their loop, they also need to override `__setattr__`

since these are brand new structs, should we instead offer per-use case structs that contain only the fields we need? 

for example: 
- a BatchProgress that has ready/started/processed/completed + a BatchLoopProgress which has total + current BatchProgress members
- an OptimizationProgress which has processed steps + OptimizationLoopProgress which has total + current OptimizationProgress members?
- an EpochProgress which has started + completed epoch counts

Comment (carmocca) [reply pytorch_lightning/trainer/progress.py:50]: > indicating this API behavior change based purely on types feels is not intuitive.

I agree it's not the clearest but to me seems like an okay tradeoff to avoid the code duplication.

> this also means if anyone extends this class

I don't think there is a reason to extend the `Progress` class in particular. `BaseProgress` and so on might but they don't contain this "unused" logic.

> since these are brand new structs, should we instead offer per-use case structs that contain only the fields we need?

> would it be easier to introduce a separate optimization progress struct which doesn't carry all these fields, and if there is demand for them later, then we can unify using the same struct?

I'm open to it, but do you have an idea how? Because this wouldn't work well:

```python
@dataclass
class Progress:  # name?
    ready: int = 0
    started: int = 0
    processed: int = 0
    completed: int = 0


@dataclass
class Progress1:  # name?
    ready: int = 0
    started: int = 0
    completed: int = 0


@dataclass
class Progress2:  # name?
    ready: int = 0
    completed: int = 0

@dataclass
class OptimizationProgress:
    optimizer: BaseProgress = BaseProgress(total=Progress1(), current=Progress1())
    scheduler: BaseProgress = BaseProgress(total=Progress2(), current=Progress2())
    zero_grad: BaseProgress = BaseProgress(total=Progress1(), current=Progress1())
```

There are several issues with this structure: 

- Duplicated docstrings between classes
- Duplicated helper functions between classes
- Typing complaints as there is no shared class
- BaseProgress assumes the 4 attributes are available

> do we think optimization progress would use these fields later?
Maybe. I couldn't think of a use for them but they could be useful in the future.

**Review by tchaton — APPROVED** [review]
Looks great ! 

I think the design is pretty elegant to avoid code duplication and improve code readability.

@ananthsub Let's get this in, so we can actually start working on the refactor. Right now, this PR is blocking us.

**Review by ananthsub — APPROVED** [review]
My preference is to revisit this design if/when we have a cleaner way to separate the tracker structs to avoid using Optional like this.

However since we don't have that now, accepting to unblock

File: pytorch_lightning/trainer/progress.py [hunk 6]
```diff
@@ -87,3 +116,65 @@ def increment_epoch_completed(self) -> None:
...
+    def optimizer_steps(self) -> int:
+        return self.optimizer.total.completed
+
+    @property
+    def scheduler_steps(self) -> int:
+        return self.scheduler.total.completed
+
+
+@dataclass
+class TrainProgress(Progress):
```
Comment (awaelchli) [inline pytorch_lightning/trainer/progress.py:144]: I vote for Training* over Train*

not blocking

**Review by awaelchli — APPROVED** [review]
